src/kvp.rs

```diff
@@ -205,6 +205,66 @@ impl EmitKVPLayer {
         let encoded_kvp_flattened: Vec<u8> = encoded_kvp.concat();
         self.send_event(encoded_kvp_flattened);
     }
+
+    /// Processes and sends a health report event based on the provided `health_report` field.
+    ///
+    /// This function extracts the `reason` field (if present)
+    /// and constructs a properly formatted health report key-value pair (KVP) using
+    /// [`build_health_kvp`]. The resulting KVP is then encoded and sent via `send_event()`.
+    ///
+    /// This function is called exclusively from [`on_event`] when a `health_report`
+    /// flag is detected, ensuring that health-related events are processed separately
+    /// from standard event logging.
+    ///
+    /// # Parameters
+    /// - `event`: The original tracing event containing metadata.
+    /// - `health_report`: The health report status as a string (e.g., `"Success"` or `"Failure"`).
+    /// - `timestamp`: The event timestamp as a formatted string.
+    fn handle_health_report(
+        &self,
+        event: &tracing::Event<'_>,
+        health_report: String,
+        event_time: Duration,
+    ) {
+        let mut reason = None;
+        event.record(
+            &mut |field: &tracing::field::Field,
+                  value: &dyn std::fmt::Debug| {
+                if field.name() == "reason" {
+                    reason = Some(
+                        format!("{:?}", value).trim_matches('"').to_string(),
+                    );
+                }
+            },
+        );
+        let timestamp = DateTime::<Utc>::from(UNIX_EPOCH + event_time)
+            .format("%Y-%m-%dT%H:%M:%S%.3f")
+            .to_string();
+
+        let agent = format!("Azure-Init/{}", env!("CARGO_PKG_VERSION"));
+        let kvp_result: Result<(String, String), String> =
+            match health_report.as_str() {
+                "success" => Ok(build_success_health_report(
+                    &agent,
+                    Some("optional_value"),
+                    &self.vm_id,
+                    &timestamp,
+                )),
+                "failure" => Ok(build_failure_health_report(
+                    &agent,
+                    reason.as_deref().unwrap_or("No failure reason provided"),
+                    Some("stuff"),
+                    &self.vm_id,
+                    &timestamp,
+                )),
+                other => Err(format!("Invalid health report kind: {}", other)),
+            };
+
+        let (key, value) =
+            kvp_result.expect("Failed to build health key-value pair");
```

Comment:
If I understand correctly, this is a case of programmer error where they used a particular key in a trace without the expected value. Given that, I think panicking might be okay assuming it'll definitely happen in the test suite, although I would include some help in the panic so the developer knows what to do to fix their oversight, something like "health_report value must be 'success' or 'failure'".